Migrate advanced external storage backends to new registration API [part 3]

RobinMcCorkell opened the pull request:
This PR migrates the following external storage backends:
- OpenStack Swift Object Storage/Rackspace
- SMB_OC -> merge with SMB
- SFTP_Key -> merge with SFTP

~~:warning: Requires the JS changes from #18440 for the SFTP_Key changes, no merge yet. Also backend unit tests need fixing, they still assume SMB_OC and SFTP_Key exist. :warning:~~

Please review @PVince81 @DeepDiver1975 @MorrisJobke @icewind1991 

cc @jmaciasportela for an example of how registration works.

Replaces #18245

Commit:
Migrate Swift external storage to new API

The Rackspace/OpenStack differences have been split into separate auth
mechanisms, with correct legacy migration

Commit:
Migrate SMB_OC external storage to new API

SMB_OC has been merged with SMB, via the identifier aliases mechanism.
Legacy migration is done to the Session Credentials password mechanism

Commit:
Migrate SFTP_Key external storage to new API

The SFTP backend now supports public key authentication alongside
password authentication.

Commit:
AjaxController uses RSA auth mechanism

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/400cb2b2-34be-4ff8-bd27-75816da06e30) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/78921cd2-5490-4b11-a9bb-97c263bd6bc5) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/01faa020-69d4-4d16-82c3-185ef18ef8ae) was created.

Commit:
Implement more fine-grained external storage permissions model

VisibilityTrait -> PermissionsTrait

PermissionsTrait stores two sets of data, $permissions and
$allowedPermissions (analogous to $visibility and $allowedVisibility of
VisibilityTrait). Each set is a map of user type ('admin' or 'personal')
to permissions (mounting permission, create permission).

The result is that a backend can now be restricted for creation, while
still allowing it to be mounted. This is useful for deprecating backends
or auth mechanisms, preventing new storages being created, while still
allowing existing storages to be mounted.

Commit:
Validate permissions for created admin storages, auth mechanism

Backend and auth mechanism permissions are checked on storage creation,
both for personal storages and for admin storages

Commit:
Prevent creation of deprecated backends

Commit:
Introduce MODIFY permission for external storages

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f969f619-4ebf-4b6c-a0c8-28d508d575bd) was created.

Xenopathic commented:
The visibility stuff has been augmented to a proper permissions model, similar to UNIX ACLs. It allows the deprecated storages to still be mounted, but not to be created (by a user or by the admin).

Xenopathic commented:
@owncloud-bot retest this please

icewind1991 commented:
smb lost the "username as share" option

icewind1991 commented:
code looks good otherwise, havent tested

Xenopathic commented:
@icewind1991 SMB never had the 'Username as share' option, only SMB_OC (which still has it, for existing storages). It can be replicated by putting '$user' in the share field though.

Xenopathic commented:
@owncloud-bot retest this please

Xenopathic commented:
@owncloud-bot retest this please

PVince81 commented:
@owncloud-bot retest this please

PVince81 commented:
Code looks good, needs testing

Xenopathic commented:
@owncloud-bot retest this please

Commit:
Fix typo in SMB_OC

Commit:
Fix Swift legacy auth mechanism fallback

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/f5b86a47-ed9a-48c1-8a00-c985a0e4508f) was created.

Xenopathic commented:
Good thing I tested the legacy migration... All 3 backends now correctly migrate to the new system - OpenStack with the 'API Key' set becomes an OpenStack with Rackspace authentication, OpenStack without 'API Key' becomes OpenStack with OpenStack authentication, SMB_OC is migrated to SMB with Session credentials (the typo was with 'Username as share', which now also works correctly). SFTP_Key correctly migrates to SFTP with RSA Public Key authentication, and it works both before and after. Generation of RSA keys works too with the new authentication mechanism.

Please review @PVince81 @MorrisJobke @nickvergessen @DeepDiver1975 @icewind1991

PVince81 commented on apps/files_external/lib/permissionstrait.php:
```diff
@@ -0,0 +1,164 @@
+ * For user type constants, see BackendService::USER_*
+ * For permission constants, see BackendService::PERMISSION_*
+ */
+trait PermissionsTrait {
```
Why do we need such complex permissions ?

Xenopathic replied:
Mainly for deprecation capabilities. In the current example, we need a way to stop users creating SMB_OC (it's deprecated), but still mount and modify existing ones. Then, the admin might want to disable mounting certain storages, so that's removing the mount and create permissions.

Basically I thought: UNIX extended permissions are well designed, so lets just emulate that.

karlitschek commented:
looks good :+1:

icewind1991 commented on apps/files_external/lib/auth/publickey/rsa.php:
```diff
@@ -0,0 +1,80 @@
+	}
+
+	public function manipulateStorageConfig(StorageConfig &$storage) {
+		$auth = new RSACrypt();
```
inject this?

Xenopathic replied:
We need a different RSACrypt object for each storage, so injection wouldn't work (there's also no Factory available)

icewind1991 commented:
:+1: otherwise

Xenopathic commented:
Core and common external tests pass, and SFTP tests pass (except for two failures with SFTP, but they also occur on master). Everything works when tested manually, and two reviewers -> merge
